DoS on ancestry POST causes SIGSEGV

### Description of Problem / Feature Request

When a large number of connections (around 1000) are made to Clair with each performing a POST on the ancestry url (when the database is being populated with test data for example), Clair will terminate with a panic (SIGSEGV)

### Expected Outcome

Clair should handle multiple connections gracefully.

### Actual Outcome

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x77c526]

goroutine 78546 [running]:
github.com/coreos/clair/database.FindAncestryAndRollback(0xc0f240, 0xc000142cd0, 0xc0014ed280, 0x19, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /data/git/va/go/src/github.com/coreos/clair/database/dbutil.go:143 +0x86
github.com/coreos/clair.IsAncestryCached(0xc0f240, 0xc000142cd0, 0xc0014ed280, 0x19, 0xc0042e56e0, 0x2, 0x2, 0xb2f7ff, 0xfca1d16886aa84b2, 0xc0042d32c0)
    /data/git/va/go/src/github.com/coreos/clair/ancestry.go:326 +0x83
github.com/coreos/clair/api/v3.(*AncestryServer).PostAncestry(0xc00013fc40, 0x7f8e0803cf60, 0xc000127a10, 0xc002710500, 0xc00013fc40, 0xc165e0, 0xc0042d32c0)
    /data/git/va/go/src/github.com/coreos/clair/api/v3/rpc.go:79 +0x185
github.com/coreos/clair/api/v3/clairpb._AncestryService_PostAncestry_Handler.func1(0x7f8e0803cf60, 0xc000127a10, 0xaaec80, 0xc002710500, 0xc000296050, 0xc00370ef60, 0xc0000c6b98, 0xc0029d0270)
    /data/git/va/go/src/github.com/coreos/clair/api/v3/clairpb/clair.pb.go:912 +0x86
github.com/coreos/clair/vendor/github.com/grpc-ecosystem/go-grpc-prometheus.UnaryServerInterceptor(0x7f8e0803cf60, 0xc000127a10, 0xaaec80, 0xc002710500, 0xc00370ef40, 0xc00370ef60, 0xc0000c6c28, 0x40e208, 0x50, 0xad44c0)
    /data/git/va/go/src/github.com/coreos/clair/vendor/github.com/grpc-ecosystem/go-grpc-prometheus/server.go:29 +0xb5
github.com/coreos/clair/api/v3/clairpb._AncestryService_PostAncestry_Handler(0xa66520, 0xc00013fc40, 0x7f8e0803cf60, 0xc000127a10, 0xc000296000, 0xb46008, 0x0, 0x0, 0x4172aa, 0x7f8e08f4bf00)
    /data/git/va/go/src/github.com/coreos/clair/api/v3/clairpb/clair.pb.go:914 +0x158
github.com/coreos/clair/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc000175900, 0xc178a0, 0xc001c16dc0, 0xc003523a00, 0xc000229410, 0x10a20d8, 0xc0029d0270, 0x0, 0x0)
    /data/git/va/go/src/github.com/coreos/clair/vendor/google.golang.org/grpc/server.go:791 +0x94c
github.com/coreos/clair/vendor/google.golang.org/grpc.(*Server).handleStream(0xc000175900, 0xc178a0, 0xc001c16dc0, 0xc003523a00, 0xc0029d0270)
    /data/git/va/go/src/github.com/coreos/clair/vendor/google.golang.org/grpc/server.go:991 +0x12b5
github.com/coreos/clair/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc001b2eb90, 0xc000175900, 0xc178a0, 0xc001c16dc0, 0xc003523a00)
    /data/git/va/go/src/github.com/coreos/clair/vendor/google.golang.org/grpc/server.go:561 +0x9f
created by github.com/coreos/clair/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
```

### Environment

- Clair version/image: latest master branch
- Clair client name/version: latest master branch
- Host OS: Ubuntu
- Kernel (e.g. `uname -a`): 4.18.0-25-generic
- Kubernetes version (use `kubectl version`): 1.13.6
- Helm version (use `helm version`): 2.14.1
- Network/Firewall setup: Vyatta

--- a/database/dbutil.go
+++ b/database/dbutil.go
@@ -140,11 +140,10 @@ func PersistNamespacesAndCommit(datastore Datastore, namespaces []Namespace) err
 // rollback.
 func FindAncestryAndRollback(datastore Datastore, name string) (Ancestry, bool, error) {
 	tx, err := datastore.Begin()
-	defer tx.Rollback()
-
 	if err != nil {
 		return Ancestry{}, false, err
 	}
+	defer tx.Rollback()
 
 	return tx.FindAncestry(name)
 }

--- a/database/dbutil_test.go
+++ b/database/dbutil_test.go
@@ -1,0 +1,33 @@
+// Copyright 2019 clair authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTxCreationFailure(t *testing.T) {
+	mdb := &MockDatastore{}
+
+	mdb.FctBegin = func() (Session, error) {
+		return nil, errors.New("Deliberately failed Begin")
+	}
+
+	_, _, err := FindAncestryAndRollback(mdb, "mockAncestry")
+	assert.Error(t, err)
+}

